torch/_dynamo/repro/after_aot.py

```diff
@@ -162,6 +180,173 @@ def deferred_for_real_inputs(real_inputs):
     return debug_wrapper
 
 
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#                       REPRO SUPPORT CODE
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+
+def _stride_or_default(stride, *, shape):
+    return stride if stride is not None else utils.make_contiguous_strides_for(shape)
+
+
+def _dtype_or_default(dtype):
+    return dtype if dtype is not None else torch.float32
+
+
+def _device_or_default(device):
+    return device if device is not None else torch.device("cpu")
+
+
+def _storage_offset_or_default(storage_offset):
+    return storage_offset if storage_offset is not None else 0
+
+
+# TODO: Support bundling the entire repro into a zip file for ease of
+# transferring around
+class InputReader:
+    def __init__(self, save_dir=None):
+        # If None, we will generate random data instead
+        if save_dir is None:
+            log.warning("no save_dir specified, will generate random data")
```

Comment:
Odd default, and I don't see this invoked w/o a save_dir, not directly nor through writer. May just remove and assert we have a save dir?

Reply:
It's important to support the no save_dir case, as this lets you run the repro script without having the real tensors.